### PR TITLE
Fix memory leak in xla::SlowOperationAlarm

### PR DESCRIPTION
**Pull request by rbaranchuk-capgemini**
Currently TSL threads (`tsl::Thread` and derived classes) can only be joined and it's done automatically in the destructor.

There's some threads that are essentially infinite loops (`xla::SlowOperationAlarm::AlarmLoop()` is an example), and as such these threads can't be joined (hence destructors for such thread objects can't be called)

Current way of dealing with this is to forget about the pointer returned from `tsl::Env::Default()->StartThread` (as is done for `xla::SlowOperationAlarm::AlarmLoop()` thread) but this approach produces memory leaks that are detected by AddressSanitizer. At the moment at least 75 tests from `//tensorflow/compiler/xla/tests:` fail due to exactly that cause.

The proposed solution is to add possibility to detach threads, and to join the thread in the destructor only if it's not detached. This also supports the design idea that every thread created should be either joined or detached and should not be left dangled.

### CODE REVIEW

**Comment by cheshire** [comment]
@joker-eph

**Comment by rbaranchuk-capgemini** [comment]
@fcoUnda

**Comment by cheshire** [comment]
>  but this approach produces memory leaks that are detected by AddressSanitizer

Could you add some more context? To the best of my knowledge, XLA tests are asan-clean.

File: tensorflow/tsl/platform/windows/env.cc [hunk 1]
```diff
@@ -53,17 +53,30 @@ class StdThread : public Thread {
  public:
   // thread_options is ignored.
   StdThread(const ThreadOptions& thread_options, const string& name,
-            std::function<void()> fn)
```
Comment (cheshire) [inline tensorflow/tsl/platform/windows/env.cc:56]: Are these changes necessary for this PR? It would be easier for me to review without them, I'm not a Windows expert.

File: tensorflow/tsl/platform/env.h [hunk 2]
```diff
@@ -570,9 +570,12 @@ class Thread {
  public:
   Thread() {}
 
-  /// Blocks until the thread of control stops running.
+  /// If a thread is not detached blocks until the thread of control stops running.
```
Comment (cheshire) [inline tensorflow/tsl/platform/env.h:573]: "it blocks"?

File: tensorflow/tsl/platform/default/env.cc [hunk 3]
```diff
@@ -76,7 +76,16 @@ class PThread : public Thread {
...
   }
 
-  ~PThread() override { pthread_join(thread_, nullptr); }
+  ~PThread() override {
+    if (!detached_) {
+      pthread_join(thread_, nullptr);
+    }
+  }
+
+  void detach() override {
```
Comment (cheshire) [inline tensorflow/tsl/platform/default/env.cc:85]: Could you write a docstring and maybe a link to some standard reference on how detached threads should behave? Is it possible to test this?

**Review by cheshire — CHANGES_REQUESTED** [review]

**Comment by rbaranchuk-capgemini** [comment]
> > but this approach produces memory leaks that are detected by AddressSanitizer
> 
> Could you add some more context? To the best of my knowledge, XLA tests are asan-clean.

It might be because they're tagged as [noasan], but it might be something else. I can run and reproduce fails locally with Clang Docker build. @fcoUnda Could you please provide more details regarding these tests and the spreadsheet they're in?

File: tensorflow/tsl/platform/windows/env.cc (hunk 1 above)
Comment (rbaranchuk-capgemini) [reply tensorflow/tsl/platform/windows/env.cc:56]: Yes, they're needed, when `std::thread` is detached it's not possible to get the id of the thread like it's currently done in the destructor `std::thread::id thread_id = thread_.get_id();` and then remove the entry from the map `GetThreadNameRegistry().erase(thread_id);`

I think it's better to make it similar to the POSIX variant, when adding/removing from the map is done in a separate function that encapsulates a call to the `fn`

File: tensorflow/tsl/platform/default/env.cc (hunk 3 above)
Comment (rbaranchuk-capgemini) [reply tensorflow/tsl/platform/default/env.cc:91]: Description added

File: tensorflow/tsl/platform/env.h (hunk 2 above)
Comment (rbaranchuk-capgemini) [reply tensorflow/tsl/platform/env.h]: The line has been corrected

**Comment by joker-eph** [comment]
Would it be possible to avoid having "infinite loop" threads, keep track of these tasks and signal them that they need to break their infinite loop and terminate properly?

**Comment by rbaranchuk-capgemini** [comment]
> Would it be possible to avoid having "infinite loop" threads, keep track of these tasks and signal them that they need to break their infinite loop and terminate properly?

That would be my suggestion too but I'm afraid I can't answer on that. I don't know particularly why there's an infinite thread in XLA and if it's possible to make it non-infinite

**Comment by fcoUnda** [comment]
@cheshire @rbaranchuk-capgemini to give a little context, there are a number of tests in tensorflow that have one or more sanitizers disabled, when they shouldn't, i.e. noasan or nomsan tags. This PR is part of an effort to analyze/fix whatever is necessary to re-enable them. @cheshire feel free to reach out internally  if you want more details.

**Comment by cheshire** [comment]
> there are a number of tests in tensorflow that have one or more sanitizers disabled, when they shouldn't, i.e. noasan or nomsan tags

Sure, but then I don't think this patch would solve it given that otherwise all XLA tests would have to have noasan?

We've discussed it a bit more, why not instead write

```
static Thread* t = tsl::Env::Default()->StartThread(....)
```

I think this should be enough to silence asan, and the resource usage would be the same?

**Comment by rbaranchuk-capgemini** [comment]
> I think this should be enough to silence asan, and the resource usage would be the same?

You're right, this works too. I'll change to this approach then.

**Review by cheshire — APPROVED** [review]

**Comment by cheshire** [comment]
Actually then we are blocked by a compiler warning:

```
unused variable 't' [-Werror,-Wunused-variable]
    static tsl::Thread* t = tsl::Env::Default()->StartThread(
                        ^
1 error generated.
```

Maybe assign it to underscore or do (void) t?

**Comment by rbaranchuk-capgemini** [comment]
> Actually then we are blocked by a compiler warning:
> 
> ```
> unused variable 't' [-Werror,-Wunused-variable]
>     static tsl::Thread* t = tsl::Env::Default()->StartThread(
>                         ^
> 1 error generated.
> ```
> 
> Maybe assign it to underscore or do (void) t?

I've added `[[maybe_unused]]`, hope it helps

**Review by cheshire — APPROVED** [review]
